### ISSUE TITLE
Subsetter drops nameIDs that make font unusable on OS X

### ISSUE THREAD

**Issue by o-t-w** [fonttools/fonttools#1170]
I am trying to add a font to Font Book (the Mac application).
It is displaying errors. When I install the font before subsetting, there are no errors. 
<img width="590" alt="screen shot 2018-02-06 at 15 53 51" src="https://user-images.githubusercontent.com/29547922/35869097-369f4f26-0b56-11e8-9adf-5ee958f9b620.png">

The font in question is the variable version of Source Sans Pro

**Comment by miguelsousa** [comment]
@o-t-w are you trying to subset a variable font? I don't think that's supported yet.

**Comment by o-t-w** [comment]
Yep, I’m trying to subset the variable version of source sans. Source Sans is gigantic without subsetting, wiping out the benefits of using a variable font :(

**Comment by miguelsousa** [comment]
Using the complete Source Sans variable fonts is still a smaller payload than using all of the individual fonts, so it's not a complete wipe out :grin:

The only workaround you have right now is to build the variable font(s) from source, subsetting the master OTFs before they get merged as a variable font. It's not that difficult; I've done it with other projects.

It seems that you're already familiar with **pyftsubset**, so all you have to do is to add a command like this to [buildVFs.sh](https://github.com/adobe-fonts/source-sans-pro/blob/master/buildVFs.sh),
```sh
# subset Roman master OTFs
masters=`find Roman/Masters/master_* -type f | egrep -i '\.otf$' | sed 's:^\./::'` # sed removes the trailing './'
for otf in $masters
do
	echo 'Subsetting' $otf
	pyftsubset $otf --unicodes=20,21,24,26,2c-2e,30-3b,3f,41-5a,61-7a,a3,2019,20ac --notdef-outline --name-IDs='*' --layout-features+=case,dnom,lnum,numr,onum,ordn,pnum,sinf,subs,sups,tnum,zero --layout-features-=locl --drop-tables-=BASE --no-subset-tables+=BASE
	mv $(echo $otf | sed 's/.otf/.subset.otf/') $otf
done
```
This chunk should be placed after **buildMasterOTFs** and before **buildCFF2VF**.

**Comment by o-t-w** [comment]
Thanks for the info Miguel! I will try that. As a front-end developer rather than a typographer, I am curious what those tables are and do. How hazardous is it to use a font with a busted sfnt and name table?
There is a [popular tutorial](https://medium.com/clear-left-thinking/how-to-use-variable-fonts-in-the-real-world-e6d73065a604) that recommends subsetting variable fonts , so I think a lot of people will try already.

**Comment by o-t-w** [comment]
Are there plans to support variable fonts?

**Comment by justvanrossum** [comment]
Subsetting Var TTF shouldn't be too hard I reckon. What else is there to do besides subsetting the `gvar` table?

**Comment by behdad** [comment]
> Subsetting Var TTF shouldn't be too hard I reckon. What else is there to do besides subsetting the gvar table?

gvar is trivial. In fact, it's already implemented.  What's not implemented right now is GDEF VarStore I think, which should be easy to get going without pruning it of unused entries.

**Comment by behdad** [comment]
Actually, I take that back.  I just tested and subsetting variable fonts works.  The VarStore in GDEF is not trimmed so it's large, but otherwise it works:
```
behdad:fonttools 0 (master)$ ./fonttools ttx -l NotoSansArabic-VF.subset.ttf 
Listing table info for "NotoSansArabic-VF.subset.ttf":
    tag     checksum   length   offset
    ----  ----------  -------  -------
    GDEF  0x02E7683D   110527     3648
    GPOS  0x44764C75       32   114176
    GSUB  0xE2F1EBC8      148   114208
    MVAR  0xDCD88C8D      180   114356
    OS/2  0x883C5B66       96     1788
    avar  0x2A7D4002       48   115024
    cmap  0x0C7C0CED       72     1884
    fvar  0x6E78DEFF      488   114536
    glyf  0x47B8A608     1268      284
    gvar  0xB1383411     7444   115072
    head  0x10C9200B       54     1624
    hhea  0x0DC903F4       36     1752
    hmtx  0x25220202       72     1680
    loca  0x0A720931       38     1584
    maxp  0x003F02E9       32     1552
    name  0xE08E168B     1658     1956
    post  0xFF9F0032       32     3616

behdad:fonttools 0 (master)$ 
```
Your error seems to be something else. Can you elaborate?

**Comment by behdad** [comment]
It might be that we drop psname during subsetting by default? Though I thought we fixed that.

**Comment by o-t-w** [comment]
I ran the command `pyftsubset SourceSansPro-var.ttf --unicodes="U+0020-007F,U+00A,U+00A9,U+00BB,U+00A0"`
Trying to add to Font Book displayed the errors. 
As a test I tried adding the variable version of source sans without subsetting and saw no errors in Font Book

**Comment by behdad** [comment]
Right. But what if you do with a non-variable version of the font? Does that work? At any rate, this seems to be a bug, NOT a missing feature. It's expected to work.

**Comment by khaledhosny** [comment]
It is complaining about `name` table, and there is a thread on typedrawers with the exact same error for non-variable font. My guess is that some of the name IDs that the subsetter drops by default are required here, so `--nsme-IDs='*'` might do the trick.

**Comment by o-t-w** [comment]
I tried running some non-variable fonts throught pyftsubset and I'm getting the same error.

**Comment by o-t-w** [comment]
--name-IDs='*' resolves the issue

**Comment by behdad** [comment]
Right. As I said, I think it's the postscriptname.

Try with `--name-IDs+=6`. If didn't work, try with +=4 as well.

**Comment by o-t-w** [comment]
Excuse my ignorance - I wish i knew more about font files. Could you explain why `--name-IDs+=6` is better than `--name-IDs='*'`

**Comment by behdad** [comment]
`--name-IDs='*'` retains all nameIDs. That might be what you want.  I'm interested in figuring out what's causing the rejectiom nto change the default to be the minimum acceptable.  Thanks.

**Comment by behdad** [comment]
Maybe @anthrotype can test that on a Mac.

**Comment by miguelsousa** [comment]
@behdad I've just tested. Font Book stops complaining if the font contains nameIDs 3, 4 and 6, in addition to what gets added by default (1, 2, IDs referenced by STAT table).

Here's the full command that makes a subset of Source Sans Variable valid according to Font Book:
`pyftsubset SourceSansVariable-Roman.ttf --unicodes="U+0020-007F,U+00A,U+00A9,U+00BB,U+00A0" --name-IDs+=3,4,6`

The original font is at https://github.com/adobe-fonts/source-sans-pro/releases/tag/variable-fonts

**Comment by RoelN** [comment]
@behdad Sorry to comment on a closed issue, am I correct to understand the only thing missing from a fully functional `pyftsubset` for variable fonts is cleaning up the VarStore in GDEF? Is that on the roadmap or shall I open an issue?

**Comment by behdad** [comment]
> @behdad Sorry to comment on a closed issue, am I correct to understand the only thing missing from a fully functional pyftsubset for variable fonts is cleaning up the VarStore in GDEF? Is that on the roadmap or shall I open an issue?
> 
> Edit: it looks like VarStore is nicely cleaned up when using pyftsubset with FontTools 3.21.3.dev0, so I might have misunderstood your comment. It appears subsetting variable fonts fully works?

I'm sure VarStore is NOT cleaned up.  So, yes, that's left. I think that's the only thing.  Wait. Checking the code, we don't subset HVAR/VVAR either.  That's trivial to add.

Yes, open an issue for finishing subsetting of varfonts and I'll try to do it.

**Comment by anthrotype** [comment]
about the new `--name-ids` default value, note that the `--help` is still saying we only keep only 1 and 2.

By re-reading these related issues, I think the consensus was to keep all the basic name ids from 1 to 6, including 5 (which wasn't included with commit 43ea974)

https://github.com/fonttools/fonttools/issues/114
https://github.com/fonttools/fonttools/issues/605

**Comment by behdad** [comment]
Thanks. Fixed. Added 0 as well.

**Comment by brawer** [comment]
I’ve done some more testing on MacOS 10.13.3.

* If a font contains only Windows names, it works perfectly fine and passes all validation tests on MacOS.

* However, if a font contains _any_ Macintosh names, at least name IDs 1, 2, 3, and 6 must be present for the Macintosh platform for validation to succeed.

Just mentioning for later reference.

**Comment by anthrotype** [comment]
thanks sascha. shall we keep this open then?
the subsetter defaults to dropping all mac legacy names. Maybe we just need to document this somewhere (in the subsetter --help probably)
